upstairs/src/lib.rs

```diff
@@ -5721,164 +5722,161 @@ impl Upstairs {
-                    failed_collate = true;
-                    error!(self.log, "Failed collate with {}", e);
-                }
-            }
-            repair_commands = ds.reconcile_task_list.len();
-        }
-
-        if failed_collate {
-            // We failed to collate the three downstairs, so we need
-            // to reset that activation request, and kick all the downstairs
-            // to FailedRepair
-            //
-            self.set_inactive(CrucibleError::RegionAssembleError).await;
-            let _active = self.active.lock().await;
-            let mut ds = self.downstairs.lock().await;
-
-            // While collating, downstairs should all be DsState::Repair.
-            // As we have released then locked the downstairs, we have to
-            // verify that the downstairs are all in the state we expect them
-            // to be.  Any change means that downstairs went away, but any
-            // downstairs still repairing should be moved to failed repair.
-            assert_eq!(ds.ds_active.len(), 0);
-            assert_eq!(ds.reconcile_task_list.len(), 0);
-
-            for (i, s) in ds.ds_state.iter_mut().enumerate() {
-                if *s == DsState::WaitQuorum {
-                    *s = DsState::FailedRepair;
-                    warn!(
-                        self.log,
-                        "Mark {} as FAILED Collate in final check", i
-                    );
-                } else {
-                    warn!(
-                        self.log,
-                        "downstairs in state {} after failed collate", *s
-                    );
-                }
-            }
-        } else if need_repair {
-            self.do_reconciliation(
-                dst,
-                lastcast,
-                ds_reconcile_done_rx,
-                repair_commands,
-            )
-            .await?;
-
-            let mut active = self.active.lock().await;
-            let mut ds = self.downstairs.lock().await;
-            /*
-             * Now that we have completed reconciliation, we move all
-             * the downstairs to the next state.  If we fail here, it means
-             * something interrupted our repair and we have to start over.
-             *
-             * As we repaired, downstairs should all be DsState::Repair.
-             *
-             * As we have released the downstairs lock while repairing, we
-             * have to verify that the downstairs are all in the state we
-             * expect them to be.  Any change means we abort and require
-             * all downstairs to reconnect, even if repair work is finished
-             * as a disconnected DS does not yet know it is all done with
-             * work and could have its state reset to New.
-             */
-
-            assert_eq!(ds.ds_active.len(), 0);
-            assert_eq!(ds.reconcile_task_list.len(), 0);
+            self.collate_downstairs(&mut ds).await
+        };
```

Comment:
Here is where the diff just falls apart, and I would suggest not trying to make any sense of it at all.